client/consensus/babe/src/lib.rs

```diff
@@ -1355,15 +1355,43 @@ pub struct BabeBlockImport<Block: BlockT, Client, I> {
 	inner: I,
 	client: Arc<Client>,
 	epoch_changes: SharedEpochChanges<Block, Epoch>,
+	genesis_slot: Option<Slot>,
 	config: Config,
 }
 
+impl<Block: BlockT, Client, I> BabeBlockImport<Block, Client, I>
+where
+	Client: HeaderBackend<Block>,
+{
+	fn genesis_slot(&mut self) -> Option<Slot> {
+		if self.genesis_slot.is_some() {
+			return self.genesis_slot.clone()
+		}
+
+		let genesis_slot =
+			self.client.header(BlockId::Number(One::one())).ok().flatten().map(|header| {
+				let pre_digest = find_pre_digest::<Block>(&header).expect(
+				"valid babe headers must contain a predigest; header has been already verified; qed",
+		    );
+
+				pre_digest.slot()
+			});
+
+		if genesis_slot.is_some() {
+			self.genesis_slot = genesis_slot.clone();
```

Comment:
Here my only concern is that we received a genesis that is not part of the canonical chain.

If this is the case then we're setting `genesis_slot` not correctly. A couple of alternatives:
1. always read genesis slot from the runtime state (i.e. from Babe's pallet `GenesisSlot` slow...).
2. do not set this value if block 1 has not yet been finalized ( `if self.client.info().finalized_number.is_zero() { return None}`)
3. read the value from the runtime state **only if** we've not finalized the first block yet